mautic/mautic
# [UI] List toolbar - select all to perform batch action

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | 🔴 
| New feature/enhancement? (use the a.x branch)      | 🟢
| Deprecations?                          | 🔴
| BC breaks? (use the c.x branch)        | 🔴
| Automated tests included?              | 🟢 
| Related user documentation PR URL      | mautic/user-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation-new#... 
| Issue(s) addressed                     | Fixes #... 



## Description
- This branch is based off https://github.com/mautic/mautic/pull/14306
- Created a select all button that when clicked, also "selects" paginated items (as opposed to the checkall checkbox which only selects items on the rendered page)
- Created batchDeleteService which perform batch delete for batchDeleteActions in various entity controllers
- Created CannotBeDeletedInterface for models where entity cannot be deleted if there are dependencies or in the case of custom fields, the entity is "fixed".






---
### 📋 Steps to test this PR:


1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Generate dummy data for testing with ddev exec bin/console doctrine:fixtures:load
3. Go to the different entity lists and see that batch actions still work when selecting individual checkboxes and when clicking on "select all".
4. See that the select all button shows how many items there are in total
5. See that the # of selected items match the total items when you click on select all
  - ![image](https://github.com/user-attachments/assets/3b1a3263-9faf-4cd3-8d41-2501b267797f)
  - ![image](https://github.com/user-attachments/assets/1b37d376-7478-4348-8796-d55f81deab4f)
6. See that Segments that segments used in other segments cannot be deleted unless you're deleting all segments
  - Same for roles with users
  - See that you can't delete your own user account
  - See that you can't delete fixed custom fields
7. Install custom plugins Clearbit and FullContact and see that they still work for both person and company lookup
